pkg/flow/internal/controller/loader.go

```diff
@@ -564,80 +568,103 @@ func (l *Loader) OriginalGraph() *dag.Graph {
 	return l.originalGraph.Clone()
 }
 
-// EvaluateDependencies re-evaluates components which depend directly or
-// indirectly on c. EvaluateDependencies should be called whenever a component
-// updates its exports.
-//
-// The provided parentContext can be used to provide global variables and
-// functions to components. A child context will be constructed from the parent
-// to expose values of other components.
-func (l *Loader) EvaluateDependencies(c *ComponentNode) {
+// EvaluateDependencies sends components which depend directly on components in updatedNodes for evaluation to the
+// workerPool. It should be called whenever components update their exports. It is beneficial to call
+// EvaluateDependencies with a batch of components, as it will enqueue the entire batch before the worker pool starts
+// to evaluate them, resulting in smaller number of total evaluations when node updates are frequent.
+func (l *Loader) EvaluateDependencies(updatedNodes []*ComponentNode) {
+	if len(updatedNodes) == 0 {
+		return
+	}
 	tracer := l.tracer.Tracer("")
-
-	l.mut.RLock()
-	defer l.mut.RUnlock()
+	spanCtx, span := tracer.Start(context.Background(), "SubmitDependantsForEvaluation", trace.WithSpanKind(trace.SpanKindInternal))
+	span.SetAttributes(attribute.Int("originators_count", len(updatedNodes)))
+	defer span.End()
 
 	l.cm.controllerEvaluation.Set(1)
 	defer l.cm.controllerEvaluation.Set(0)
-	start := time.Now()
-
-	spanCtx, span := tracer.Start(context.Background(), "GraphEvaluatePartial", trace.WithSpanKind(trace.SpanKindInternal))
-	span.SetAttributes(attribute.String("initiator", c.NodeID()))
-	defer span.End()
 
-	logger := log.With(l.log, "trace_id", span.SpanContext().TraceID())
-	level.Info(logger).Log("msg", "starting partial graph evaluation")
-	defer func() {
-		span.SetStatus(codes.Ok, "")
+	l.mut.RLock()
+	defer l.mut.RUnlock()
 
-		duration := time.Since(start)
-		level.Info(logger).Log("msg", "finished partial graph evaluation", "duration", duration)
-		l.cm.componentEvaluationTime.Observe(duration.Seconds())
-	}()
+	dependenciesToParentsMap := make(map[dag.Node]dag.Node)
+	for _, parent := range updatedNodes {
+		// Make sure we're in-sync with the current exports of parent.
+		l.cache.CacheExports(parent.ID(), parent.Exports())
+		// We collect all nodes directly incoming to parent.
+		_ = dag.WalkIncomingNodes(l.graph, parent, func(n dag.Node) error {
+			dependenciesToParentsMap[n] = parent
+			return nil
+		})
+	}
 
-	// Make sure we're in-sync with the current exports of c.
-	l.cache.CacheExports(c.ID(), c.Exports())
+	// Submit all dependencies for asynchronous evaluation.
+	// During evaluation, if a node's exports change, Flow will add it to updated nodes queue (controller.Queue) and
+	// the Flow controller will call EvaluateDependencies on it again. This results in a concurrent breadth-first
+	// traversal of the nodes that need to be evaluated.
```

Comment:
Is there a situation in which the exports change, but we have already scheduled something that depends on those exports to run, and hence there could be a race condition where a downstream component reads a value which is more recent than the one it's supposed to read?

Reply:
Reading more recent value is okay and it could already happen currently, because `component.Queue` also accepts only one instance of a component to be queued at the time. I was able to reproduce some intermediate updated being missed in https://github.com/grafana/agent/pull/5278 - but that is okay as long as a dependency always eventually evaluates after its parent node has updated and the latest value propagates to all the necessary nodes.